minor QOL improvements to live stream

ehpersonal38 opened the pull request:
If a user starts playing a recording while the live stream is already playing, the live stream volume will be set to 0 until the recording is finished playing. Prevents overlapping audio.

Open to feedback/suggestions. Thanks!

mcguirepr89 commented:
Damn, this is cool!

mcguirepr89 commented:
I have a challenge for you that I have been putting off:

The latest changes that I had to do to accommodate RaspiOS discontinuing the default `pi` user somehow broke the Spectrogram view's refresh cycle. It used to refresh at the same interval as whatever the RECORDING_LENGTH is set to (and still does on installations that were performed prior to the changes that allow dynamic users). On my installation, however, that feature is no longer working. 

I've been putting off the troubleshooting for this since I'm working on things for the new repo, but it is bothering me!

No worries if you have other things on your plate (things for which I am a grateful beneficiary!)

Thanks again for this super neat tweak

mcguirepr89 commented:
Nevermind!!! I fixed it :)

ehpersonal38 commented:
> Nevermind!!! I fixed it :)

Wow, that's timing! I just came back to this page to tell you I fixed it! I'm interested to see what you came up with? Here's my code for spectrogram.php:

> 
if (file_exists(realpath(dirname(__FILE__)).'/thisrun.txt')) {
  $config = parse_ini_file(realpath(dirname(__FILE__)).'/thisrun.txt');
} elseif (file_exists(realpath(dirname(__FILE__)).'/thisrun.ini')) {
  $config = parse_ini_file(realpath(dirname(__FILE__)).'/thisrun.ini');
}
$refresh = $config['RECORDING_LENGTH'];
$time = time();
echo "<img style=\"width:100%;height:100%\" src=\"/spectrogram.png?nocache=$time\">";
header("Refresh: $refresh;");

And glad to help with the tweak!

mcguirepr89 commented:
In true Patrick fashion, I lumped this fix into another experiment/hunch I was working on :)

I'm learning all of this stuff as I go, so I have a lot of forms that `POST`, but the more I use and change this stuff, the more inclined I am to think that using `GET` instead would allow me to share some values among the PHP scripts a little easier.

SOOO, I changed ALLL `POST`s to `GET`s, and changed `homepage/index.html` to always only show the `views.php` in the `iframe` instead of weird rules for `spectrogram.php` and viewing the log. Now, the spectrogram page and the log both have the navigation bar at the top, which will help with keeping the live stream going while navigating the pages.

Also, I noticed, as you did!!!, that `parse_ini_file` was not finding the new location for `thisrun.txt` -- your solution seems more elegant than mine, wherein I just hardcoded it (again) -- yours checks the running script's directory for the file?

I really appreciate it!!!

Take a look on my tester to see how it performs now: https://virginia.birdnetpi.com

You'll see the commits come through after a little more testing :)

ehpersonal38 commented:
I'm glad you were able to get it working while improving some other areas along the way! GETs and POSTs have always confused me, and compounded with these scripts working inside an iframe puts me deeply outside of what I'm used to working with, but I think I'm slowly getting a handle of it.

On the topic of that, I put together a little mockup of how spectogram's refresh can be accomplished in javascript, which obviates the need for the iframe to refresh (looks cleaner & uses a little less bandwidth). Just giving you some ideas in case you ever find a similar need for something like this!

```php
<?php
if (file_exists(realpath(dirname(__FILE__)).'/thisrun.txt')) {
  $config = parse_ini_file(realpath(dirname(__FILE__)).'/thisrun.txt');
} elseif (file_exists(realpath(dirname(__FILE__)).'/thisrun.ini')) {
  $config = parse_ini_file(realpath(dirname(__FILE__)).'/thisrun.ini');
}
$refresh = $config['RECORDING_LENGTH'];
echo "<img id='spectrogramimage' style=\"width:100%;height:100%\" src=\"/spectrogram.png?nocache=$time\">";
?>

<script>
// every $refresh seconds, this loop will run and refresh the spectrogram image
window.setInterval(function(){
  document.getElementById("spectrogramimage").src = "/spectrogram.png?nocache="+Date.now();
}, <?php echo $refresh; ?>*1000);
</script>
```

mcguirepr89 commented:
I love this idea! I will use it here for the spectrogram and also on `overview.php` to update the chart and "Currently Analyzing" :)

Thanks for all the goodies! I'm loving being able to add so many cool improvements in one day
